src/backend/src/signer.rs

```diff
@@ -151,7 +159,52 @@ pub async fn btc_principal_to_p2wpkh_address(
 ///
 /// # Errors
 /// Errors are enumerated by: `TopUpCyclesLedgerError`
-#[allow(clippy::unused_async)] // TODO: Remove once the code is implemented
 pub async fn top_up_cycles_ledger(request: TopUpCyclesLedgerRequest) -> TopUpCyclesLedgerResult {
-    todo!("Add code that tops up the cycles ledger per this request: {request:?}")
+    let cycles_ledger = CyclesLedgerService(*CYCLES_LEDGER);
+    let account = Account {
+        owner: ic_cdk::id(),
+        subaccount: None,
+    };
+    let (ledger_balance,): (Nat,) = cycles_ledger
+        .icrc_1_balance_of(&account)
+        .await
+        .map_err(|_| TopUpCyclesLedgerError::CouldNotGetBalanceFromCyclesLedger)?;
+
+    if ledger_balance < request.threshold() {
```

Comment:
Not for this PR but, realizing now: should we have two thresholds?

- One threshold for the ledger balance, as in this line

but also

- Another threshold for the backend canister cycles itself under which we do not want to topup the account?